src/data/load_geometry.js

```diff
@@ -48,40 +47,57 @@ function resample(ring) {
  */
 export default function loadGeometry(feature: VectorTileFeature, canonical): Array<Array<Point>> {
     if (!canonical) return [];
+
     const cs = projection.tileTransform(canonical);
-    const reproject = (p, featureExtent) => {
-        const s = Math.pow(2, canonical.z)
-        const x_ = (canonical.x + p.x / featureExtent) / s;
-        const y_ = (canonical.y + p.y / featureExtent) / s;
-        const l = new MercatorCoordinate(x_, y_).toLngLat();
-        const {x, y} = projection.project(l.lng, l.lat);
-        p.x = (x * cs.scale - cs.x) * EXTENT;
-        p.y = (y * cs.scale - cs.y) * EXTENT;
-    };
-    const scale = EXTENT / EXTENT;
+    const z2 = Math.pow(2, canonical.z);
+    const featureExtent = feature.extent;
+
+    function reproject(p) {
+        const lng = lngFromMercatorX((canonical.x + p.x / featureExtent) / z2);
+        const lat = latFromMercatorY((canonical.y + p.y / featureExtent) / z2);
+        const {x, y} = projection.project(lng, lat);
+        return new Point(
+            (x * cs.scale - cs.x) * EXTENT,
+            (y * cs.scale - cs.y) * EXTENT
+        );
+    }
+
+    function addResampled(resampled, startMerc, endMerc, startProj, endProj) {
```

Comment:
It was initially a closure because it captures `canonical`, `featureExtent` and `cs` (tile transform), all of which would have to be moved to arguments in a quite a verbose way if we move it out. We could do it but it's also not a bottleneck.